src/python_bindings/py_util/logging.cpp

```diff
@@ -0,0 +1,122 @@
+#include <memory>
+#include <mutex>
+#include <string>
+#include <string_view>
+
+#include <pybind11/pybind11.h>
+#include <pybind11/stl.h>
+#include <spdlog/sinks/base_sink.h>
+#include <spdlog/spdlog.h>
+
+#include "util/logger.h"
+
+namespace {
+namespace py = pybind11;
+// A global flag set by the Python atexit handler to signal that the
+// interpreter is shutting down. The logging sink checks this flag to prevent
+// unsafe calls back into the Python C-API during this critical phase.
+bool g_is_shutting_down = false;
+
+constexpr int kPyTraceLevel = 5;
+
+template <typename Mutex>
+class PythonLoggerSink : public spdlog::sinks::base_sink<Mutex> {
+public:
+    explicit PythonLoggerSink(py::object py_logger) : py_logger_(std::move(py_logger)) {}
+
+protected:
+    void sink_it_(spdlog::details::log_msg const& msg) override {
+        py::gil_scoped_acquire gil;
+
+        if (g_is_shutting_down || !py_logger_ || py_logger_.is_none()) {
+            return;
+        }
+
+        int msg_py_level;
+        switch (msg.level) {
+            case spdlog::level::trace:
+                msg_py_level = kPyTraceLevel;
+                break;
+            case spdlog::level::debug:
+                msg_py_level = 10;
+                break;
+            case spdlog::level::info:
+                msg_py_level = 20;
+                break;
+            case spdlog::level::warn:
+                msg_py_level = 30;
+                break;
+            case spdlog::level::err:
+                msg_py_level = 40;
+                break;
+            case spdlog::level::critical:
+                msg_py_level = 50;
+                break;
+            default:
+                msg_py_level = 0;
+                break;
+        }
+
+        if (!py_logger_.attr("isEnabledFor")(msg_py_level).template cast<bool>()) {
+            return;
+        }
+
+        std::string_view payload(msg.payload.data(), msg.payload.size());
+        (void)py_logger_.attr("log")(msg_py_level, payload);
+    }
+
+    void flush_() override {
+        py::gil_scoped_acquire gil;
+        if (g_is_shutting_down || !py_logger_ || py_logger_.is_none()) {
+            return;
+        }
+        (void)py_logger_.attr("handlers").attr("flush")();
```

Comment:
`.handlers` is a `list`, this doesn't work. Why would we do that in the first place? The handlers are Python's domain, I don't see a reason to mess with them.